### PR TITLE
fix: KMS access denied errors when downloading the static env file

### PR DESCRIPTION
**Pull request by bencehornak**
## Bug description
Since Copilot 1.32.0 cross account deployments are broken for all services (except static website) due to lacking permissions on the task execution role to perform the `kms:Decrypt` operation on the required key.

## Background
Since #5329 the artifacts are encrypted with a [Kms Key](https://github.com/bencehornak/copilot-cli/blob/1146bf995d687bfff615a5c7d35266fb8bcb0b56/internal/pkg/template/templates/app/cf.yml#L17-L63), which belongs to the App stack.

Due to the wrong configuration of the `ExecutionRole` resource if one deploys the workload to a different account than where the app stack is located, the download of the env file fails due to the lacking `kms:Decrypt` permission on the corresponding key.

https://github.com/bencehornak/copilot-cli/blob/67db7a2d4f94c69c836586f714341e587b1b8220/internal/pkg/template/templates/workloads/partials/cf/executionrole.yml#L39-L43

Note that this policy grants access to all keys in the _workload's_ account (which is not a great practice either, see #4628), which is wrong, considering that the key is located in the app's account.

## Fix
The policy was changed to the following:
```
- Effect: 'Allow'
  Action:
    - 'kms:Decrypt'
  Resource:
    - !Ref ArtifactKeyARN
```

where `ArtifactKeyARN` is the ARN of the Kms Key.

## TODO
- [x] There are many failing test cases, because the generated YAML changes. Is there a way to efficiently fix all of these test cases?
- [x] Would it make sense to add a test case, which prevents this bug in the feature? If yes, how? -> changed some test cases to test if the KMS key ARN gets passed to the stacks

## Conclusion

Fixes partially: #4628.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

**Comment by ghost** [comment]
This workaround worked for fixing deployments for a Backend Service with Copilot 1.32.0:

```yaml
- op: replace
  path: /Resources/ExecutionRole/Properties/Policies/0/PolicyDocument/Statement/2/Resource/0
  value: arn:aws:kms:....:key/... # the KMS key's ARN, which you see exported on the app's CloudFormation stack
```

**Comment by bencehornak** [comment]
This line might also be buggy
https://github.com/aws/copilot-cli/blob/cce61ba2c0bfee8939a4bcd51a46ea602d1e87ba/internal/pkg/template/templates/workloads/partials/cf/instancerole.yml#L60

UPDATE: yes, it was, fixed it in a later commit.

File: internal/pkg/deploy/cloudformation/stack/testdata/workloads/backend/http-autoscaling-params.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
 		"LogRetention": "30",
 		"ContainerPort": "8080",
 		"EnvFileARN": "",
+		"ArtifactKeyARN": "",
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/testdata/workloads/backend/http-autoscaling-params.json]: Can we in at least one integ test verifying `ArtifactKeyARN` can be rendered correctly with the app stackset resource info?

**Review by iamhopaul123 — APPROVED** [review]
Awesome! Thank you for fixing this bug!

**Review by iamhopaul123 — COMMENTED** [review]
It would be great if you can fix another wrong reference for the KMS key [here](https://github.com/aws/copilot-cli/blob/c6aefafeee89b5a52b56c4f25d6f14013516d115/internal/pkg/template/templates/workloads/services/static-site/cf.yml#L334) for the static site. We can have the same `ArtifactKeyARN` param for consistency and address https://github.com/aws/copilot-cli/issues/5472 🙏

**Comment by bencehornak** [comment]
Sure, will address your points tomorrow, @iamhopaul123

**Comment by bencehornak** [comment]
@iamhopaul123 I believe I have resolved both of your remarks, see my last 2 commits

**Comment by iamhopaul123** [comment]
> @iamhopaul123 I believe I have resolved both of your remarks, see my last 2 commits

Looks good! Thank you!

**Review by CaptainCarpensir — APPROVED** [review]
Changes look good to me! Thanks so much for contributing ❤️

**Pull request merged**